### ISSUE TITLE
[Vulkan/Raspberry Pi 4] Vulkan subsystem 13 not supported

### ISSUE THREAD

**Issue by vanfanel** [hrydgard/ppsspp#14672]
### What happens?
When I run my PPSSPP build (aarch64) on a Raspberry Pi 4, after selecting the Vulkan video backend, I get a message saying:
`Vulkan subsystem 13 not supported`

OpenGL_ES2 video backend seems to work.

### What *should* happen?
Vulkan video backend should work, I guess.

### What are you using?
Raspberry Pi 4, Debian 10 aarch64, latest MESA 21.1.5, latest libSDL2 built from latest GIT sources.
Vulkan test in SDL2 works. Quake 3 and vkQuake work. RetroArch using Vulkan backend works.

#### What graphics card (GPU) or mobile phone model?
Raspberry Pi 4 (VideoCore, which uses the V3D driver in MESA).

#### What PPSSPP version (standalone/official), and did it work before?
First time trying Vulkan backend on PPSSPP on the Pi4 (maybe I am the first person to try with latest MESA, etc)

#### Which game or games?
No games. PPSSPP wont even load its GUI.

#### Checklist

 - [X ] Test in the [latest git build](https://buildbot.orphis.net/ppsspp/) in case it's already fixed.
 - [ X] [Search for other reports](https://github.com/hrydgard/ppsspp/search?q=my+issue&type=issues) of the same issue.
 - [X ] Try resetting settings or older versions and include if the issue is related.
 - [X ] Include logs or screenshots of issue.

NOTE: I opened this issue to find out if PPSSPP works against current MESA on the Pi4, because I would like to report it to the V3D driver devs if its lacking something that PPSSPP needs.

**Comment by unknownbrackets** [comment]
I think that means SDL_SYSWM_KMSDRM, which may mean that #13135 is now fixed.  I think calls to vkCreateDisplayPlaneSurfaceKHR need to be implemented for this to work.

-[Unknown]

**Comment by vanfanel** [comment]
@unknownbrackets 

Regarding https://github.com/hrydgard/ppsspp/issues/13135, yes, it should be fixed now because, as I said, PPSSPP is now working on SDL2 KMS/DRM using GLES2.

As for Vulkan, the vkCreateDisplayPlaneSurfaceKHR() function on SDL2 KMS/DRM backend is being called here:
https://github.com/libsdl-org/SDL/blob/d4e1b4974ac8711cc89bd187c9f412f84771372b/src/video/kmsdrm/SDL_kmsdrmvulkan.c#L432

The vkCreateDisplayPlaneSurfaceKHR() is implemented on the Pi4, and it works perfectly well with ONE limitation: no windows, so no arbitrary window sizes, only sizes with a matching physical resolutions are supported.
Other than that, or works, and Vulkan programs in general work on SDL2 without X11.

**Comment by unknownbrackets** [comment]
Sounds like we should be using `SDL_Vulkan_CreateSurface()` then.  I think it didn't exist yet in SDL when we added Vulkan support.  We're actually manually implementing the surface creation per window system right now.

-[Unknown]

**Comment by vanfanel** [comment]
@unknownbrackets Ah! Now I fully understand the problem. Since you use SDL2, I assumed you were already hitting SDL_Vulkan_CreateSurface().

Yes, using `SDL_Vulkan_CreateSurface()` is a very easy solution and you would get instant Vulkan support on a variety of graphic stacks, including Wayland, KMS/DRM, etc... without the hassle of manual Vulkan surface creation via `vkCreateDisplayPlaneSurfaceKHR()`

But I must say I had to go through some hack-ish solutions to get Vulkan working on the KMS/DRM backend in SDL2, since Vulkan uses the KMS/DRM subsystems under the hood... So manual `vkCreateDisplayPlaneSurfaceKHR()` could also be considered.

Unless you want to get Vulkan on both KMSDRM and Wayland support, too. As I said, with `SDL_Vulkan_CreateSurface()` you would get support for all of them with a single solution.

**Comment by orbea** [comment]
Seems a similar issue is on destop linux too. https://github.com/hrydgard/ppsspp/issues/13135#issuecomment-891011648

**Comment by vanfanel** [comment]
@orbea That issue is unrelated. This is an issue affecting Vulkan initialization only. OpenGL on KMSDRM on SDL2 on a Pi4 works nowadays, just use current SDL2 GIT version.

**Comment by orbea** [comment]
Read the linked comment, I meant the vulkan issue.

**Comment by vanfanel** [comment]
I have implemented the specific extension code, but I need someone with good Vulkan knowledge to build my fork and help me to look for the remaining problem:

https://github.com/vanfanel/ppsspp.git

**Comment by retropieuser** [comment]
Is there any news on this? Or it's a wider problem? As it affects the pi 5 and I suspect with a switch to Vulkan you'd see a good performance boost given the pi 5's better spec

**Comment by akamming** [comment]
i have the same issue (on the same rpi4 with Vulkan), does anyone know how to get around this issue?

**Comment by amverweij** [comment]
The libretro core does work with vulkan.

**Comment by akamming** [comment]
Yeah tx, sorry and i agree. I should be more specific: inknip it works on libretto core, but I would like the native ppsspp for better performance. So I am looking for a fix or workaround on the native ppsspp

**Comment by retropieuser** [comment]
Do you use RetroPie? Opening the game with XINIT:/ on RetroPie opens it as an x11 window and allows Vulkan to load

**Comment by akamming** [comment]
I am indeed using retropie. How can I configure this?

**Comment by retropieuser** [comment]
Not sure which version you're on, you might need to update or manually install the whole of retropie for 64bit rather than the standard 32bit version that is on the website.

But assuming you're on 64bit and have updated run command you can add this to psp emulators.cfg

ppsspp-dev = "XINIT-WM:/opt/retropie/emulators/ppsspp-dev; /opt/retropie/emulators/ppsspp-dev/PPSSPPSDL --fullscreen %ROM%; popd"

I'm using the ppsspp-dev version (which just builds a newer branch) which you might need for Vulkan support, I'm not 100% sure

The important bit is the XINIT-WM: this tells run command to open an x11 window

**Comment by akamming** [comment]
A check, I understand now. But I already have it working under X11. So this would not add anything. I want to run without X11. My use case is a selfbuilt arcade cabinet. So i can’t have it boot into X11.
